lib/autoupdate/installer.go

```diff
@@ -0,0 +1,302 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package autoupdate
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"text/template"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	checksumType   = "sha256"
+	checksumHexLen = 64
+)
+
+// LocalAgentInstaller manages the creation and removal of installations
+// of Teleport.
+type LocalAgentInstaller struct {
+	// InstallDir contains each installation, named by version.
+	InstallDir string
+	// HTTP is an HTTP client for downloading Teleport.
+	HTTP *http.Client
+	// Log contains a logger.
+	Log *slog.Logger
+	// ReservedFreeTmpDisk is the amount of disk that must remain free in /tmp
+	ReservedFreeTmpDisk uint64
+	// ReservedFreeInstallDisk is the amount of disk that must remain free in the install directory.
+	ReservedFreeInstallDisk uint64
+}
+
+// Remove a Teleport version directory from InstallDir.
+func (ai *LocalAgentInstaller) Remove(ctx context.Context, version string) error {
+	versionDir := filepath.Join(ai.InstallDir, version)
+	sumPath := filepath.Join(versionDir, checksumType)
+
+	// invalidate checksum first, to protect against partially-removed
+	// directory with valid checksum.
+	if err := os.Remove(sumPath); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := os.RemoveAll(versionDir); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Install a Teleport version directory in InstallDir.
+func (ai *LocalAgentInstaller) Install(ctx context.Context, version, template string) error {
+	versionDir := filepath.Join(ai.InstallDir, version)
+	sumPath := filepath.Join(versionDir, checksumType)
+
+	// generate download URI from template
+	uri, err := makeURL(template, version)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Get new and old checksums. If they match, skip download.
+	// Otherwise, clear the old version directory and re-download.
+	checksumURI := uri + "." + checksumType
+	newSum, err := ai.getChecksum(ctx, checksumURI)
+	if err != nil {
+		return trace.Errorf("failed to download checksum from %s: %w", checksumURI, err)
+	}
+	oldSum, err := readChecksum(sumPath)
+	if err == nil {
+		if bytes.Equal(oldSum, newSum) {
+			ai.Log.InfoContext(ctx, "Version already present.", "version", version)
+			return nil
+		}
+		ai.Log.WarnContext(ctx, "Removing version that does not match checksum.", "version", version)
+		if err := ai.Remove(ctx, version); err != nil {
+			return trace.Wrap(err)
+		}
+	} else if !errors.Is(err, os.ErrNotExist) {
```

Comment:
should we handle the other type of error?

Reply:
This handles all errors besides ErrNotExist